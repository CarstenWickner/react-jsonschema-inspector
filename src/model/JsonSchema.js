import RefScope from "./RefScope";

/**
 * Representation of a Json Schema, offering a number of convenient functions for traversing and extracting information.
 */
export default class JsonSchema {
    /**
     * Raw JSON Schema.
     *
     * @type {object}
     */
    schema;

    /**
     * Configuration steering how the json schema is being traversed/parsed.
     *
     * @type {{oneOf: ?{groupTitle: ?string}, anyOf: ?{groupTitle: ?string}}}
     */
    parserConfig;

    /**
     * @type {RefScope}
     */
    scope;

    /**
     * Constructor for a JsonSchema (wrapper).
     *
     * @param {object} schema - the JSON Schema to represent
     * @param {object} parserConfig - configuration affecting how the json schema is being traversed/parsed
     * @param {?RefScope} scope - collection of available $ref targets (will be generated based on `schema` if not provided)
     */
    constructor(schema, parserConfig, scope) {
        this.schema = schema;
        this.parserConfig = parserConfig;
        this.scope = scope || new RefScope(this);
    }
}
