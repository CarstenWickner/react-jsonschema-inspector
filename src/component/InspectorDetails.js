import PropTypes from "prop-types";
import React from "react";

import InspectorDetailsContent from "./InspectorDetailsContent";

import JsonSchema from "../model/JsonSchema";

const InspectorDetails = ({
    columnData, renderSelectionDetails, renderEmptyDetails
}) => {
    const lastColumnContainsSelection = columnData.length && columnData[columnData.length - 1].trailingSelection;
    const selectionColumnIndex = columnData.length - (lastColumnContainsSelection ? 1 : 2);
    const trailingSelectionColumn = selectionColumnIndex < 0 ? null : columnData[selectionColumnIndex];
    const itemSchema = trailingSelectionColumn ? trailingSelectionColumn.items[trailingSelectionColumn.selectedItem] : null;
    return (
        <div className="jsonschema-inspector-details">
            {itemSchema && renderSelectionDetails && renderSelectionDetails({
                itemSchema,
                selectionColumnIndex,
                columnData
            })}
            {itemSchema && !renderSelectionDetails && (
                <InspectorDetailsContent
                    itemSchema={itemSchema}
                    selectionColumnIndex={selectionColumnIndex}
                    columnData={columnData}
                />
            )}
            {!itemSchema && renderEmptyDetails && renderEmptyDetails({
                rootColumnSchemas: columnData.length ? columnData[0].items : {}
            })}
        </div>
    );
};

InspectorDetails.propTypes = {
    columnData: PropTypes.arrayOf(PropTypes.shape({
        items: PropTypes.objectOf(PropTypes.instanceOf(JsonSchema)).isRequired,
        selectedItem: PropTypes.string,
        trailingSelection: PropTypes.bool
    })).isRequired,
    /** func({ itemSchema: JsonSchema, selectionColumnIndex: number, columnData }) */
    renderSelectionDetails: PropTypes.func,
    /** func({ rootColumnSchemas }) */
    renderEmptyDetails: PropTypes.func
};

InspectorDetails.defaultProps = {
    renderSelectionDetails: null,
    renderEmptyDetails: null
};

export default InspectorDetails;
