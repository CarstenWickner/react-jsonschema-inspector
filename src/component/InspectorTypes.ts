import * as React from "react";

import { JsonSchema } from "../model/JsonSchema";
import { JsonSchemaGroup } from "../model/JsonSchemaGroup";

import { RawJsonSchema, KeysOfRawJsonSchemaStringValues } from "../types/RawJsonSchema";
import { ParserConfig } from "../types/ParserConfig";
import { RenderOptions } from "../types/RenderOptions";

export interface InspectorDefaultProps {
    /**
     * Array of additional JSON Schemas that may be referenced by entries in `schemas` but are not shown (on the root level) themselves.
     */
    referenceSchemas?: Array<RawJsonSchema>;
    /**
     * Flag indicating whether the root column should be hidden in case of `schemas` containing only a single item.
     */
    hideSingleRootItem?: boolean;
    /**
     * Array of (default) selected items – each item representing the selection in one displayed column.
     */
    defaultSelectedItems?: Array<string | Array<number>>;
    /**
     * Options for the traversing/parsing of JSON schemas. Defining how optional parts of a schema should be represented.
     */
    parserConfig?: ParserConfig;
    /**
     * Function accepting a `JsonSchema` instance representing an array's declared type of items and returning an object listing the available
     * properties to offer. The default, providing access to the array's items, is: `arrayItemSchema => ({ "[0]": arrayItemSchema })`
     * Provided inputs are (1) a `JsonSchema` representing the declared type of array items, (2) a `JsonSchemaGroup` representing the surrounding
     * array's type, (3) an `Array.<number>` indicating the 'optionIndexes' selected in the array (i.e. #2), which may be undefined.
     * The expected values of the expected object being returned may be either `JsonSchema`, plain/raw json schema definitions, or a mix of them.
     * E.g. `arrayItemSchema => ({ "[0]": arrayItemSchema, "length": { type: "number" } })` is valid here as well.
     */
    buildArrayProperties?: (
        // declared type of the array's items
        schema: JsonSchema,
        // schema group representing the array
        schemaGroup: JsonSchemaGroup,
        // selected optionIndexes in the array's JsonSchemaGroup (if it contains options)
        optionIndexes?: Array<number>
    ) => Record<string, JsonSchema | RawJsonSchema>;
    /**
     * Options for the breadcrumbs feature shown in the footer – set to `null` to turn it off.
     * - "prefix": Text to show in front of root level selection, e.g. "//" or "./"
     * - "separator": Text to add between the selected item names from adjacent columns, e.g. "." or "/"
     * - "skipSeparator": Function to identify breadcrumb names that should not be prepended with a "separator"
     * - "mutateName": Function to derive the selected item's representation in the breadcrumbs from their name
     * - "preventNavigation": Flag indicating whether double-clicking an item should preserve subsequent selections, otherwise they are discarded
     * - "renderItem": Custom render function for a single breadcrumb item
     * - "renderTrailingContent": Custom render function for adding extra elements (e.g. a "Copy to Clipboard" button) after the breadcrumbs
     */
    breadcrumbs?: null | {
        /*
         * Text to show in front of root level selection, e.g. "//" or "./"
         */
        prefix?: string;
        /*
         * Text to add between the selected item names from adjacent columns, e.g. "." or "/"
         */
        separator?: string;
        /*
         * Function to identify breadcrumb names that should not be prepended with a "separator"
         */
        skipSeparator?: (name: string, column: RenderColumn, index: number) => boolean;
        /*
         * Function to derive the selected item's representation in the breadcrumbs from their name
         */
        mutateName?: (selectedItem: string, column: RenderColumn, index: number) => undefined | null | string;
        /*
         * Flag indicating whether double-clicking an item should preserve subsequent selections, otherwise they are discarded
         */
        preventNavigation?: boolean;
        /*
         * Custom render function for a single breadcrumb item.
         */
        renderItem?: (props: { breadcrumbText: string; hasNestedItems: boolean; column: RenderColumn; index: number }) => React.ReactElement;
        /*
         * Custom render function for adding extra elements (e.g. a "Copy to Clipboard" button) after the breadcrumbs.
         */
        renderTrailingContent?: (props: { breadcrumbTexts: Array<string>; columnData: Array<RenderColumn> }) => React.ReactElement;
    };
    /**
     * Options for the search input shown in the header and its impact on the displayed columns – set to `null` to turn it off.
     * - "byPropertyName": Flag indicating whether property names should be considered when searching/filtering
     * - "fields": Array of strings: each referring to a textual field in a JSON Schema (e.g. `["title", "description"]`) in which to search
     * - "filterBy": Custom search/filter logic, if present: overriding behaviour based on "fields"
     * - "inputPlaceholder": Hint text to display in the search input field (defaults to "Search")
     * - "debounceWait": Number indicating the delay in milliseconds since the last change to the search term before applying it. Default: `200`.
     * - "debounceMaxWait": Number indicating the maximum delay in milliseconds before a newly entered search is being applied. Default: `500`.
     */
    searchOptions?: null | {
        byPropertyName?: boolean;
        fields?: Array<KeysOfRawJsonSchemaStringValues>;
        filterBy?: (enteredSearchFilter: string | null) => (rawSchema: RawJsonSchema) => boolean | undefined;
        inputPlaceholder?: string;
        debounceWait?: number;
        debounceMaxWait?: number;
    };
    /**
     * Callback to invoke after the selection changed.
     * Expects two inputs:
     * 1. the string-array of selected items
     * 2. object containing a "columnData" key, holding the full render information for all columns (except for currently applied search/filter)
     */
    onSelect?: (
        newSelection: Array<string | Array<number>>,
        newRenderData: { columnData: Array<RenderColumn> },
        breadcrumbsTexts?: Array<string>
    ) => void;
    /**
     * Custom render function for the header tool-bar (to the left of the search field).
     */
    renderHeaderToolBar?: (props: { columnData: Array<RenderColumn> }) => React.ReactElement;
    /**
     * Custom render function for the content of a single item in a column.
     * Expects a single object as input with the following keys:
     * - "name": providing the name of the respective item
     * - "hasNestedItems": flag indicating whether selecting this item may display another column with further options to the right
     * - "selected": flag indicating whether the item is currently selected
     * - "schemaGroup": the full `JsonSchemaGroup` associated with the item
     */
    renderItemContent?: (props: {
        name: string;
        hasNestedItems: boolean;
        selected: boolean;
        schemaGroup: JsonSchemaGroup;
        optionIndexes?: Array<number>;
    }) => React.ReactElement;
    /**
     * Custom render function for the details block on the right (only used if there is an actual selection).
     * Expects a single object as input with the following keys:
     * - "itemSchemaGroup": the full `JsonSchemaGroup` associated with the currently selected trailing item (i.e. right-most selection)
     * - "columnData": the full render information for all columns
     * - "selectionColumnIndex": indicating the index of the right-most column containing a selected item (for convenient use of "columnData")
     * - "optionIndexes": in case of an optional path being selected
     */
    renderSelectionDetails?: (props: {
        itemSchemaGroup: JsonSchemaGroup;
        columnData: Array<RenderColumn>;
        selectionColumnIndex: number;
        optionIndexes?: Array<number>;
    }) => React.ReactElement;
    /**
     * Custom render function for the details block on the right (only used if there is no selection).
     * Expects a single object as input with the following key:
     * - "rootColumnSchemas": the full render information for the root column (since there is no selection, there are no other columns)
     */
    renderEmptyDetails?: (props: { rootColumnSchemas: Record<string, JsonSchemaGroup> }) => React.ReactElement;
}

export interface InspectorProps extends InspectorDefaultProps {
    /**
     * Object containing names of root level items (as keys) each associated with their respective JSON Schema (as values).
     */
    schemas: Record<string, RawJsonSchema>;
}

interface RenderColumnDetails {
    trailingSelection?: boolean;
    onSelect: (event: React.SyntheticEvent, selectedItem?: string | Array<number>) => void;
}

export interface RenderItemsColumn extends RenderColumnDetails {
    items: Record<string, JsonSchemaGroup>;
    selectedItem?: string;
    filteredItems?: Array<string>;
}

export interface RenderOptionsColumn extends RenderColumnDetails {
    options: RenderOptions;
    contextGroup: JsonSchemaGroup;
    selectedItem?: Array<number>;
    filteredItems?: Array<Array<number>>;
}

export type RenderColumn = RenderItemsColumn | RenderOptionsColumn;
