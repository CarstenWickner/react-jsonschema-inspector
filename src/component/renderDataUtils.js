import PropTypes from "prop-types";

import JsonSchema from "../model/JsonSchema";
import JsonSchemaGroup from "../model/JsonSchemaGroup";
import {
    createGroupFromSchema, getPropertiesFromSchemaGroup, getTypeOfArrayItemsFromSchemaGroup,
    createOptionTargetArrayFromIndexes, getOptionsInSchemaGroup, getIndexPermutationsForOptions
} from "../model/schemaUtils";
import { isDefined, isNonEmptyObject, mapObjectValues } from "../model/utils";
import { createFilterFunctionForSchema } from "../model/searchUtils";

function isOptionIndexValidForOptions(optionIndexes, options) {
    let optionsPart = options;
    optionIndexes.forEach((index) => {
        if (optionsPart && optionsPart.options && optionsPart.options.length > index) {
            optionsPart = optionsPart.options[index];
        } else {
            optionsPart = null;
        }
    });
    return optionsPart && !optionsPart.options;
}

function createRootColumnData(schemas, referenceSchemas, parserConfig) {
    // first prepare those schemas that may be referenced by the displayed ones or each other
    const referenceScopes = [];
    referenceSchemas.forEach((rawRefSchema) => {
        const newReferenceScope = new JsonSchema(rawRefSchema, parserConfig).scope;
        referenceScopes.forEach((otherReferenceScope) => {
            newReferenceScope.addOtherScope(otherReferenceScope);
            otherReferenceScope.addOtherScope(newReferenceScope);
        });
        referenceScopes.push(newReferenceScope);
    });
    // the first column always lists all top-level schemas
    const rootColumnItems = mapObjectValues(schemas, (rawSchema) => {
        const schema = new JsonSchema(rawSchema, parserConfig);
        referenceScopes.forEach(referenceScope => schema.scope.addOtherScope(referenceScope));
        return createGroupFromSchema(schema);
    });
    return { items: rootColumnItems };
}

function buildDefaultArrayProperties(arrayItemSchema) {
    return { "[0]": arrayItemSchema };
}

function buildNextColumn(schemaGroup, optionIndexes, buildArrayProperties = buildDefaultArrayProperties) {
    if (!optionIndexes) {
        const options = getOptionsInSchemaGroup(schemaGroup);
        if (options.options) {
            // next column should offer the selection of options within the schema group
            return {
                contextGroup: schemaGroup,
                options
            };
        }
    }
    // next column should list all available properties
    const propertySchemas = getPropertiesFromSchemaGroup(schemaGroup, optionIndexes);
    if (isNonEmptyObject(propertySchemas)) {
        // convert the individual JsonSchema values into JsonSchemaGroups
        return {
            items: mapObjectValues(propertySchemas, createGroupFromSchema)
        };
    }
    // there are no properties, so this might be an array
    const nestedArrayItemSchema = getTypeOfArrayItemsFromSchemaGroup(schemaGroup, optionIndexes);
    if (nestedArrayItemSchema) {
        // next column should allow accessing the schema of the array's items
        const arrayProperties = mapObjectValues(
            buildArrayProperties(nestedArrayItemSchema),
            propertyValue => (
                propertyValue instanceof JsonSchema ? propertyValue : new JsonSchema(propertyValue, nestedArrayItemSchema.parserConfig)
            )
        );

        return {
            items: mapObjectValues(arrayProperties, createGroupFromSchema)
        };
    }
    return {};
}

export function createRenderDataBuilder(onSelectInColumn) {
    return (schemas, referenceSchemas, selectedItems, parserConfig, buildArrayProperties) => {
        // the first column always lists all top-level schemas
        let nextColumn = createRootColumnData(schemas, referenceSchemas, parserConfig);
        let selectedSchemaGroup;
        const columnData = selectedItems.map((selection, index) => {
            const currentColumn = nextColumn;
            const isOptionSelection = typeof selection !== "string";
            let isValidSelection;
            if (isOptionSelection && selectedSchemaGroup && currentColumn.options) {
                isValidSelection = isOptionIndexValidForOptions(selection, currentColumn.options);
            } else if (!isOptionSelection && currentColumn.items) {
                selectedSchemaGroup = currentColumn.items[selection];
                isValidSelection = isDefined(selectedSchemaGroup);
            }
            if (isValidSelection) {
                nextColumn = buildNextColumn(selectedSchemaGroup, isOptionSelection ? selection : undefined, buildArrayProperties);
                if (isOptionSelection) {
                    selectedSchemaGroup = null;
                }
            } else {
                nextColumn = {};
            }
            // name of the selected item (i.e. key in 'items') or int array of option indexes
            currentColumn.selectedItem = isValidSelection ? selection : null;
            currentColumn.onSelect = onSelectInColumn(index);
            return currentColumn;
        }).filter(({ items, options }) => items || options);
        // set the flag for the last column containing a valid selection
        const columnCount = columnData.length;
        if (columnCount) {
            // there is at least one column, check whether the last column has a valid selection
            const selectedItemInLastColumn = columnData[columnCount - 1].selectedItem;
            // if the last column has no valid selection, the second to last column must have one
            if (selectedItemInLastColumn || columnCount > 1) {
                // there is at least one column with a valid selection, mark the column with the trailing selection as such
                columnData[selectedItemInLastColumn ? (columnCount - 1) : (columnCount - 2)].trailingSelection = true;
            }
        }
        // append last column where there is no selection yet, unless the last selected item has no nested items or options of its own
        if (isNonEmptyObject(nextColumn)) {
            nextColumn.onSelect = onSelectInColumn(selectedItems.length);
            columnData.push(nextColumn);
        }
        // wrap the result into a new object in order to make this more easily extensible in the future
        return { columnData };
    };
}

/**
 * Check whether a given JsonSchema has any nested properties or is an array.
 *
 * @param {JsonSchema} schema single schema instance to check in (ignoring any nested sub schemas, e.g. `allOf`/`oneOf`/`anyOf`)
 * @return {Boolean} return whether any properties are mentioned or `schema` represents an array with items of a certain type
 */
function hasSchemaNestedItems(schema) {
    const { schema: rawSchema } = schema;
    const { properties, required, items, additionalItems } = rawSchema;
    return isNonEmptyObject(properties)
        || required
        || isNonEmptyObject(items)
        || isNonEmptyObject(additionalItems);
}

/**
 * Check whether a given JsonSchemaGroup has any nested properties, is an array or contains selectable optional sub-schemas.
 *
 * @param {JsonSchemaGroup} schemaGroup single schema group instance to check in
 * @return {Boolean} return whether any properties are mentioned, it represents an array with items of a certain type or contains selectable options
 */
export function hasSchemaGroupNestedItems(schemaGroup, optionIndexes) {
    return schemaGroup.someEntry(hasSchemaNestedItems, createOptionTargetArrayFromIndexes(optionIndexes))
        || (!optionIndexes && getOptionsInSchemaGroup(schemaGroup).options);
}

const optionShape = {
    groupTitle: PropTypes.string
};
optionShape.options = PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.shape({}), PropTypes.shape(optionShape)])).isRequired;

/**
 * Proptype validation of the `contextGroup` prop (and the related `options`/`items` props).
 *
 * @param {Object} param0 props
 * @param {?JsonSchemaGroup} param0.contextGroup group to validate
 * @param {?Object} param0.options representation of options within the `contextGroup`
 * @param {?Object} param0.items list of selectable entries being listed (i.e. properties/options/array-item-accessors)
 * @return {Error|null} validation error or null if validation is successful
 */
function contextGroupValidator({ contextGroup, options, items }) {
    if (isDefined(contextGroup) && !(contextGroup instanceof JsonSchemaGroup)) {
        return new Error("`contextGroup` is not a JsonSchemaGroup");
    }
    if (isNonEmptyObject(options) !== isDefined(contextGroup)) {
        return new Error("`options` and `contextGroup` are expected to be both set or not at the same time");
    }
    if (isNonEmptyObject(items) === isNonEmptyObject(options)) {
        return new Error("exactly one of `items` or `options` should be provided");
    }
    return null;
}

/**
 * PropType validation of the `selectedItem` prop.
 *
 * @param {Object} param0 props
 * @param {?String|Array.<Number>} param0.selectedItem name of the selected property/array-item-accessor or array of option indexes
 * @return {Error|null} validation error or null if validation is successful
 */
function selectedItemValidator({ selectedItem }) {
    if (isDefined(selectedItem)
        && typeof selectedItem !== "string"
        && !(
            Array.isArray(selectedItem) && selectedItem.length && selectedItem.every(arrayEntry => typeof arrayEntry === "number")
        )) {
        return new Error("`selectedItem` is not a string or array of numbers");
    }
    // assume all ok
    return null;
}

/**
 * PropType validation of the `trailingSelection` prop (and the related `selectedItem` prop).
 *
 * @param {Object} param0 props
 * @param {?String|Array.<Number>} param0.selectedItem name of the selected property/array-item-accessor or array of option indexes
 * @param {?Boolean} param0.trailingSelection flag indicating whether the `selectedItem` is the last selection (i.e. in the right-most column)
 * @return {Error|null} validation error or null if validation is successful
 */
function trailingSelectionValidator({ selectedItem, trailingSelection }) {
    if (isDefined(trailingSelection) && typeof trailingSelection !== "boolean") {
        return new Error("`trailingSelection` is not a boolean");
    }
    if (trailingSelection && !selectedItem) {
        return new Error("`trailingSelection` is true while there is no `selectedItem`");
    }
    return null;
}

/**
 * PropType validation of the `filteredItems` prop.
 *
 * @param {Object} param0 props
 * @param {?Array.<String>|Array.<Array.<Number>>} param0.filteredItems names of properties/options/array-item-accessors matchign the search filter
 * @param {?Object} param0.options representation of options within the `contextGroup`
 * @param {?Object} param0.items list of selectable entries being listed (i.e. properties/options/array-item-accessors)
 * @return {Error|null} validation error or null if validation is successful
 */
function filteredItemsValidator({ items, options, filteredItems }) {
    if (isDefined(filteredItems)) {
        if (!Array.isArray(filteredItems)) {
            return new Error("`filteredItems` is not an `array`");
        }
        if (items && filteredItems.some(singleItem => !items[singleItem])) {
            return new Error("`filteredItems` are not all part of `items`");
        }
        if (options && !filteredItems.every(singleOption => isOptionIndexValidForOptions(singleOption, options))) {
            return new Error("`filteredItems` are not all part of index combinations derived from `options`");
        }
    }
    return null;
}

/**
 * Shape for PropType validation of a single entries in the standard `columnData` array generated by a created render-data-builder.
 *
 * @param {?Boolean} includeOnSelect flag indicating whether `onSelect` should be treated as required
 * @return {Object} return `columnData` shape for PropType validation
 */
export function getColumnDataPropTypeShape(includeOnSelect = true) {
    return {
        items: PropTypes.objectOf(PropTypes.instanceOf(JsonSchemaGroup)),
        options: PropTypes.shape(optionShape),
        contextGroup: contextGroupValidator,
        selectedItem: selectedItemValidator,
        trailingSelection: trailingSelectionValidator,
        filteredItems: filteredItemsValidator,
        onSelect: includeOnSelect ? PropTypes.func.isRequired : PropTypes.func // func(SyntheticEvent: event, string|array<number>: identifier)
    };
}

/**
 * Create a function that returns the list of `filteredItems` for a given entry in the standard `columnData` array.
 *
 * @param {Function} flatSearchFilter function determining whether a given raw json schema matches some search filter
 * @param {Object} flatSearchFilter.param0 raw json schema
 * @param {Boolean} flatSearchFilter.return indication whether raw json schema is matching search filter
 * @return {Function} return function that returns the list of `filteredItems` for a given entry in the standard `columnData` array
 * @return {Object} return.param0 entry in `columnData` array to check
 * @return {?JsonSchemaGroup} return.param0.contextGroup json schema group containing options to select from
 * @return {?Object} return.param0.options representation of options within the `contextGroup`
 * @return {?Object} return.param0.items list of selectable entries being listed (i.e. properties/options/array-item-accessors)
 * @return {?Array.<String>|Array.<Array.<Number>>} return.return names of properties/options/array-item-accessors matchign the search filter
 */
export function createFilterFunctionForColumn(flatSearchFilter) {
    const containsMatchingItems = createFilterFunctionForSchema(flatSearchFilter);
    return ({ items, options, contextGroup }) => {
        if (isNonEmptyObject(items)) {
            return Object.keys(items).filter(key => items[key].someEntry(containsMatchingItems));
        }
        return getIndexPermutationsForOptions(options)
            .filter(optionIndexes => contextGroup.someEntry(containsMatchingItems, createOptionTargetArrayFromIndexes(optionIndexes)));
    };
}
