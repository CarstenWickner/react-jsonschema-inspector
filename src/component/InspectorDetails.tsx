import * as React from "react";

import { InspectorDetailsContent } from "./InspectorDetailsContent";
import { JsonSchemaGroup } from "../model/JsonSchemaGroup";
import { InspectorProps, RenderColumn, RenderItemsColumn, RenderOptionsColumn } from "./InspectorTypes";

export const InspectorDetails: React.FunctionComponent<{
    columnData: Array<RenderColumn>;
    renderSelectionDetails?: InspectorProps["renderSelectionDetails"];
    renderEmptyDetails?: InspectorProps["renderEmptyDetails"];
}> = ({ columnData, renderSelectionDetails, renderEmptyDetails }): React.ReactElement => {
    const lastColumnContainsSelection = columnData.length && columnData[columnData.length - 1].trailingSelection;
    const selectionColumnIndex = columnData.length - (lastColumnContainsSelection ? 1 : 2);
    const trailingSelectionColumn = selectionColumnIndex < 0 ? null : columnData[selectionColumnIndex];
    let itemSchemaGroup: JsonSchemaGroup | undefined = undefined;
    let optionIndexes: Array<number> | undefined = undefined;
    if (trailingSelectionColumn) {
        itemSchemaGroup = (trailingSelectionColumn as RenderItemsColumn).items
            ? (trailingSelectionColumn as RenderItemsColumn).items[trailingSelectionColumn.selectedItem as string]
            : (trailingSelectionColumn as RenderOptionsColumn).contextGroup;
        if ((trailingSelectionColumn as RenderOptionsColumn).options) {
            optionIndexes = trailingSelectionColumn.selectedItem as Array<number>;
        }
    }
    return (
        <div className="jsonschema-inspector-details">
            {itemSchemaGroup &&
                renderSelectionDetails &&
                renderSelectionDetails({
                    itemSchemaGroup,
                    selectionColumnIndex,
                    columnData,
                    optionIndexes
                })}
            {itemSchemaGroup && !renderSelectionDetails && (
                <InspectorDetailsContent itemSchemaGroup={itemSchemaGroup} selectionColumnIndex={selectionColumnIndex} columnData={columnData} />
            )}
            {!itemSchemaGroup &&
                renderEmptyDetails &&
                renderEmptyDetails({
                    rootColumnSchemas: columnData.length ? (columnData[0] as RenderItemsColumn).items : {}
                })}
        </div>
    );
};
