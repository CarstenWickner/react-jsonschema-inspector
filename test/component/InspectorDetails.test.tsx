import * as React from "react";
import { shallow } from "enzyme";

import { InspectorDetails } from "../../src/component/InspectorDetails";
import { createRenderDataBuilder } from "../../src/component/renderDataUtils";
import { JsonSchema } from "../../src/model/JsonSchema";
import { JsonSchemaGroup } from "../../src/model/JsonSchemaGroup";
import { RenderItemsColumn, RenderColumn } from "../../src/component/InspectorTypes";

describe("renders correctly", () => {
    const buildColumnData = createRenderDataBuilder(() => (): void => {});
    it("with minimal/default props", () => {
        const { columnData } = buildColumnData(
            {
                Foo: {
                    title: "Bar",
                    description: "Foobar"
                }
            },
            [],
            ["Foo"],
            {}
        );
        const component = shallow(<InspectorDetails columnData={columnData} />);
        expect(component).toMatchSnapshot();
    });
    describe("with empty columnData", () => {
        it("shows nothing by default", () => {
            const component = shallow(<InspectorDetails columnData={[]} />);
            expect(component.children().exists()).toBe(false);
        });
        it("applies custom renderEmptyDetails", () => {
            const component = shallow(
                <InspectorDetails
                    columnData={[]}
                    renderEmptyDetails={({ rootColumnSchemas }): React.ReactElement => (
                        <span className="custom-empty-details">{Object.keys(rootColumnSchemas).length}</span>
                    )}
                />
            );
            expect(component.find(".custom-empty-details").text()).toBe("0");
        });
    });
    describe("with no selection", () => {
        const { columnData } = buildColumnData(
            {
                Foo: {
                    title: "Bar",
                    description: "Foobar"
                }
            },
            [],
            [],
            {}
        );

        it("shows nothing by default", () => {
            const component = shallow(<InspectorDetails columnData={columnData} />);
            expect(component.children().exists()).toBe(false);
        });
        it("applies custom renderEmptyDetails", () => {
            const component = shallow(
                <InspectorDetails
                    columnData={columnData}
                    renderEmptyDetails={({ rootColumnSchemas }): React.ReactElement => (
                        <span className="custom-empty-details">{Object.keys(rootColumnSchemas).length}</span>
                    )}
                />
            );
            expect(component.find(".custom-empty-details").text()).toBe("1");
        });
    });
    describe("with array item selection", () => {
        const schema = {
            title: "Bar",
            items: { $ref: "#/$defs/itemSchema" },
            $defs: {
                itemSchema: {
                    description: "Array Item",
                    type: "object"
                }
            }
        };
        const { columnData: columnDataProp } = buildColumnData({ Foo: schema }, [], ["Foo", "[0]"], {});

        it("show InspectorDetailsContent by default", () => {
            const component = shallow(<InspectorDetails columnData={columnDataProp} />);
            const { columnData, itemSchemaGroup, selectionColumnIndex } = component.childAt(0).props() as {
                itemSchemaGroup: JsonSchemaGroup;
                columnData: Array<RenderColumn>;
                selectionColumnIndex: number;
            };
            expect(columnData).toEqual(columnDataProp);
            expect(itemSchemaGroup).toEqual(((columnDataProp[1] as unknown) as RenderItemsColumn).items["[0]"]);
            expect(selectionColumnIndex).toEqual(1);
        });
        it("applies custom renderSelectionDetails", () => {
            const renderSelectionDetails = jest.fn(() => <span className="custom-selection-details" />);
            const component = shallow(<InspectorDetails columnData={columnDataProp} renderSelectionDetails={renderSelectionDetails} />);
            expect(component.exists(".custom-selection-details")).toBe(true);
            expect(renderSelectionDetails.mock.calls).toHaveLength(1);
            const { columnData, itemSchemaGroup, selectionColumnIndex } = renderSelectionDetails.mock.calls[0][0] as {
                itemSchemaGroup: JsonSchemaGroup;
                columnData: Array<RenderColumn>;
                selectionColumnIndex: number;
            };
            expect(columnData).toEqual(columnDataProp);
            expect(itemSchemaGroup).toEqual(((columnDataProp[1] as unknown) as RenderItemsColumn).items["[0]"]);
            expect(selectionColumnIndex).toBe(1);
        });
    });
    describe("with option selection", () => {
        const subSchema = {
            title: "Sub Title"
        };
        const schema = {
            title: "Bar",
            oneOf: [
                subSchema,
                {
                    description: "Sub Description"
                }
            ]
        };
        const { columnData: columnDataProp } = buildColumnData({ Foo: schema }, [], ["Foo", [0]], {});

        it("show InspectorDetailsContent by default", () => {
            const component = shallow(<InspectorDetails columnData={columnDataProp} />);
            const { columnData, itemSchemaGroup, selectionColumnIndex } = component.childAt(0).props() as {
                itemSchemaGroup: JsonSchemaGroup;
                columnData: Array<RenderColumn>;
                selectionColumnIndex: number;
            };
            expect(columnData).toEqual(columnDataProp);
            expect((itemSchemaGroup.entries[0] as JsonSchema).schema).toEqual(schema);
            expect(selectionColumnIndex).toEqual(1);
        });
        it("applies custom renderSelectionDetails", () => {
            const renderSelectionDetails = jest.fn(() => <span className="custom-selection-details" />);
            const component = shallow(<InspectorDetails columnData={columnDataProp} renderSelectionDetails={renderSelectionDetails} />);
            expect(component.exists(".custom-selection-details")).toBe(true);
            expect(renderSelectionDetails.mock.calls).toHaveLength(1);
            const { columnData, itemSchemaGroup, selectionColumnIndex } = renderSelectionDetails.mock.calls[0][0] as {
                itemSchemaGroup: JsonSchemaGroup;
                columnData: Array<RenderColumn>;
                selectionColumnIndex: number;
            };
            expect(columnData).toEqual(columnDataProp);
            expect((itemSchemaGroup.entries[0] as JsonSchema).schema).toEqual(schema);
            expect(selectionColumnIndex).toBe(1);
        });
    });
});
