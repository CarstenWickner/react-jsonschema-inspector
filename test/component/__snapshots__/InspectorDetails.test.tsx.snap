// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with minimal/default props 1`] = `
<div
  className="jsonschema-inspector-details"
>
  <Component
    columnData={
      Array [
        Object {
          "items": Object {
            "Foo": JsonSchemaAllOfGroup {
              "entries": Array [
                JsonSchema {
                  "parserConfig": Object {},
                  "schema": Object {
                    "description": "Foobar",
                    "title": "Bar",
                  },
                  "scope": RefScope {
                    "externalRefs": Map {},
                    "internalRefs": Map {
                      "#" => [Circular],
                    },
                    "otherScopes": Array [],
                  },
                },
              ],
            },
          },
          "onSelect": [Function],
          "selectedItem": "Foo",
          "trailingSelection": true,
        },
      ]
    }
    itemSchemaGroup={
      JsonSchemaAllOfGroup {
        "entries": Array [
          JsonSchema {
            "parserConfig": Object {},
            "schema": Object {
              "description": "Foobar",
              "title": "Bar",
            },
            "scope": RefScope {
              "externalRefs": Map {},
              "internalRefs": Map {
                "#" => [Circular],
              },
              "otherScopes": Array [],
            },
          },
        ],
      }
    }
    selectionColumnIndex={0}
  />
</div>
`;
